File: DESIGN-CSharp/Observer/Subject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Observer
{
    public abstract class Subject
    {
        protected List<Adventure> list = new List<Adventure>();

        //想接收通知的觀察者
        public void add(Adventure observer) 
        {
            list.Add(observer);
        }

        //不想接收通知的觀察者
        public void remove(Adventure observer)
        {
            list.Remove(observer);
        }

        //發布任務
        public abstract void sendQuestions(String questions);
    }

    public class Association : Subject
    {
        public override void sendQuestions(String questions) 
        {
            foreach (var adventure in list)
            {
                adventure.getQuestions(questions);
            }
        }
    }
}

File: BehavioralPatterns/Observer/Subject.py
from abc import ABC, abstractmethod
from Adventurer import  Adventure

# 被觀察者介面
class Subject(ABC):
    def __init__(self):
        self._list = []

    def add(self, observer: Adventure):
        self._list.append(observer)

    def remove(self, observer: Adventure):
        self._list.remove(observer)
    
    @abstractmethod
    def sendQuestions(self, questions: str):
        pass

class Association(Subject):
    def sendQuestions(self, questions: str):
        for adventure in self._list:
            adventure.getQuestions(questions)